### PR TITLE
Warn when user wants to configure properties in way it is not fully supported

### PR DESCRIPTION
**Pull request by damianszczepanik**
I was quite surprised that even i switched on checkstyle.configurationFile changes were applied for all file. I though this is sputnik error but it seems to be lack of support. I added some validation so nobody won't be surprised :)

### CODE REVIEW

File: src/main/java/pl/touk/sputnik/Main.java [hunk 1]
```diff
@@ -31,8 +36,17 @@ public static void main(String[] args) {
         }
 
         ConfigurationHolder.initFromFile(commandLine.getOptionValue(CliOption.CONF.getCommandLineParam()));
-        ConfigurationHolder.instance().updateWithCliOptions(commandLine);
-        ConnectorFacade facade = ConnectorFacadeFactory.INSTANCE.build(ConfigurationHolder.instance().getProperty(GeneralOption.CONNECTOR_TYPE));
+        Configuration configuration = ConfigurationHolder.instance();
+        configuration.updateWithCliOptions(commandLine);
+
+        ConnectorType connectorType = ConnectorType.valueOf(configuration.getProperty(GeneralOption.CONNECTOR_TYPE));
```
Comment (damianszczepanik) [inline src/main/java/pl/touk/sputnik/Main.java]: Changed the extractor so it is NPE-safe and when unsupported connector is passed, default one is returned. Now it works like other methods.

File: src/main/java/pl/touk/sputnik/Main.java [hunk 2]
```diff
@@ -31,8 +36,17 @@ public static void main(String[] args) {
...
+        ConnectorFacade facade = ConnectorFacadeFactory.INSTANCE.build(connectorType);
+        try {
+            facade.supports(ConfigurationHolder.instance());
+        } catch (GerritException | GeneralOptionNotSupportedException e) {
```
Comment (damianszczepanik) [inline src/main/java/pl/touk/sputnik/Main.java]: No

File: src/main/java/pl/touk/sputnik/Main.java [hunk 3]
```diff
@@ -31,8 +36,17 @@ public static void main(String[] args) {
...
+            System.out.println(e.getMessage());
+            System.exit(1);
+        }
```
Comment (damianszczepanik) [inline src/main/java/pl/touk/sputnik/Main.java]: So it's static now - and that's not what you like.

File: src/main/java/pl/touk/sputnik/connector/ConnectorFacade.java [hunk 4]
```diff
@@ -12,5 +15,14 @@
...
     List<ReviewFile> listFiles();
 
+    /**
+     * Validates if given options are supported by selected connector. If not it throws an exception with information
+     * which option won't be supported and shall be changed.
+     * 
+     * @throws GeneralOptionNotSupportedException
+     *             if configuration is invalid
+     */
+    void supports(Configuration configuration) throws GeneralOptionNotSupportedException;
```
Comment (damianszczepanik) [inline src/main/java/pl/touk/sputnik/connector/ConnectorFacade.java]: That was your suggestion as mentioned above but I agree that validate tells more what is the purpose of the method.

File: src/main/java/pl/touk/sputnik/connector/ConnectorFacadeFactory.java [hunk 5]
```diff
@@ -1,27 +1,25 @@
...
-    private static final String GERRIT = "gerrit";
-    private static final String STASH = "stash";
-
     GerritFacadeBuilder gerritFacadeBuilder = new GerritFacadeBuilder();
     StashFacadeBuilder stashFacadeBuilder = new StashFacadeBuilder();
 
     @NotNull
-    public ConnectorFacade build(String key) {
-        switch (key) {
+    public ConnectorFacade build(ConnectorType type) {
```
Comment (damianszczepanik) [inline src/main/java/pl/touk/sputnik/connector/ConnectorFacadeFactory.java]: OK, I also changed the type of the exception thrown by this method. Previously IllegalArgumentException crashed Sputnik, now GeneralOptionNotSupportedException is supported by Main.getConnectorFacade()

File: src/main/java/pl/touk/sputnik/Main.java [hunk 6]
```diff
@@ -26,17 +30,32 @@ public static void main(String[] args) {
             commandLine = cliWrapper.parse(args);
         } catch (ParseException e) {
             printUsage(cliWrapper);
-            System.out.println(e.getMessage());
+            System.err.println(e.getMessage());
```
Comment (SpOOnman) [inline src/main/java/pl/touk/sputnik/Main.java]: Please don't change it to err. I want this message to be printed to out, even though I know it's an error.

File: src/main/java/pl/touk/sputnik/Main.java [hunk 7]
```diff
@@ -26,17 +30,32 @@ public static void main(String[] args) {
...
 
+    private static ConnectorFacade getConnectorFacade(Configuration configuration) {
+        ConnectorType connectorType = ConnectorType.getValidConnectorType(configuration
+                .getProperty(GeneralOption.CONNECTOR_TYPE));
+        ConnectorFacade facade = null;
+        try {
+            facade = ConnectorFacadeFactory.INSTANCE.build(connectorType);
+            facade.validate(ConfigurationHolder.instance());
+        } catch (GeneralOptionNotSupportedException e) {
+            System.err.println(e.getMessage());
```
Comment (SpOOnman) [inline src/main/java/pl/touk/sputnik/Main.java]: Please use out here as well.

**Comment by SpOOnman** [comment]
There's a last issue with out only.

I have one concern - what does github presents Sonar changes as new files comparing to master? I think  there should be a rebase.

Can you please squash your changes into one commit and submit a new PR? Like you did with other one? Git history would be much cleaner with this.

**Comment by damianszczepanik** [comment]
Yes, something went wrong with rebase :(
New PR #66 is ready - that would be easier for sure.

**Pull request closed without merging**